control-plane/controller/api-gateway/consul/config_entry_translation.go

```diff
@@ -0,0 +1,207 @@
+package consul
+
+import (
+	"time"
+
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	consulAPI "github.com/hashicorp/consul/api"
+)
+
+func GatewayToAPIGateway(k8sGW gwv1beta1.Gateway) consulAPI.APIGatewayConfigEntry {
+	listeners := make([]consulAPI.APIGatewayListener, 0, len(k8sGW.Spec.Listeners))
+	conditions := make([]consulAPI.Condition, 0, len(k8sGW.Status.Conditions))
+	for _, listener := range k8sGW.Spec.Listeners {
+		certificates := make([]consulAPI.ResourceReference, 0, len(listener.TLS.CertificateRefs))
+		for _, certificate := range listener.TLS.CertificateRefs {
+			c := consulAPI.ResourceReference{
+				Kind:        consulAPI.InlineCertificate,
+				Name:        string(certificate.Name),
+				SectionName: "",
+				Partition:   "",
+				Namespace:   string(*certificate.Namespace),
+			}
+			certificates = append(certificates, c)
+		}
+		l := consulAPI.APIGatewayListener{
+			Name:     string(listener.Name),
+			Hostname: string(*listener.Hostname),
+			Port:     int(listener.Port),
+			Protocol: string(listener.Protocol),
+			TLS: consulAPI.APIGatewayTLSConfiguration{
+				Certificates: certificates,
+				MaxVersion:   "",
+				MinVersion:   "",
+				CipherSuites: []string{},
+			},
+		}
+
+		listeners = append(listeners, l)
+	}
+
+	for _, condition := range k8sGW.Status.Conditions {
+		// TODO: John Maguire - convert status/reason/type to use a mapping of values defined
+		// in consul to convert from the k8s status/type/reason into consul status/types/reason
+		c := consulAPI.Condition{
+			Type:    condition.Type,
+			Status:  string(condition.Status),
+			Reason:  condition.Reason,
+			Message: condition.Message,
+			Resource: &consulAPI.ResourceReference{
+				Kind:        k8sGW.Kind,
+				Name:        k8sGW.Name,
+				SectionName: "",
+				Partition:   "",
+				Namespace:   k8sGW.Namespace,
+			},
+			LastTransitionTime: &time.Time{},
+		}
+		conditions = append(conditions, c)
+	}
+
+	for _, listener := range k8sGW.Status.Listeners {
+		for _, condition := range listener.Conditions {
+			listenerCondition := consulAPI.Condition{
+				Type:    condition.Type,
+				Status:  string(condition.Status),
+				Reason:  condition.Reason,
+				Message: condition.Message,
+				Resource: &consulAPI.ResourceReference{
+					Kind:        consulAPI.APIGateway,
+					Name:        k8sGW.Name,
+					SectionName: string(listener.Name),
+					Partition:   "",
+					Namespace:   k8sGW.Namespace,
+				},
+				LastTransitionTime: ptrTo(time.Now().UTC()),
+			}
+			conditions = append(conditions, listenerCondition)
+		}
+	}
+
+	return consulAPI.APIGatewayConfigEntry{
+		Kind:      k8sGW.Kind,
+		Name:      k8sGW.Name,
+		Meta:      map[string]string{}, // TODO: what should go in here?
```

Comment:
this is kinda my main question: what meta do we copy over from the k8s type?

Reply:
So, here's an example of what gets used for our service registrations, we'll probably want something roughly similar to mark things as managed by us and, due to the way we transform names/namespaces to/from Kubernetes, something to mark what Kubernetes objects these actually correspond to:

https://github.com/hashicorp/consul-k8s/blob/11f842cbe94b5df544c03e2441a1820397f92b9b/control-plane/connect-inject/controllers/endpoints/endpoints_controller.go#L404-L410